member_creator/member_creator.1
.\" Hey, EMACS: -*- nroff -*-
.TH MEMBER_CREATOR "1" "Aug 2014" "member_creator"
.SH NAME
member_creator \- Starship Factory membership system
.SM LDAP
account synchronization
.SH SYNOPSIS
.TP
.B member_creator
\fI--config=PATH\fR
[\fI--dry-run\fR]
[\fI--verbose\fR]
.SH DESCRIPTION
.PP
.B member_creator
is a command line utility for synchronizing the membership management system
data with an
.SM LDAP
account database.
It essentially creates new users in
.SM LDAP
and sends out welcome e\-mails for members joining the organization and purges
.SM LDAP
users (i.e. it deletes them or removes them from the relevant groups if they
are also a member of other groups) when their membership is terminated in the
web frontend.
.SH OPTIONS
.TP
.B \-\-config=PATH
indicates to
.B member_creator
where to find the configuration file with more detailed settings.
.TP
.B \-\-dry\-run
instructs
.B member_creator
to process the database records and move them to the member list as it usually
would, but not to touch
.SM LDAP
in the process.
.TP
.B \-\-verbose
instructs
.B member_creator
to disclose some additional information concerning the progress of the
membership creation or removal.
.SH SECURITY
.PP
.B member_creator
needs administrative access to an
.SM LDAP
tree in order to be able to add and remove users and modify
.SM LDAP
groups.
This in itself bears the risk of compromise because with this kind of access
it would be trivial to create new users with extended rights and known
password, and thus gain
.SM UNIX
user access to all machines using
.SM LDAP
for authentication.
.PP
It is, however, not running as a web service but more of a batch queue
processor, which limits its exposure.
.PP
.B member_creator
does not need to run with extended privileges, so it is recommended to run
it as a separate user without any special permissions.
It also doesn't need to run as the same user as the web frontend.
.PP
The configuration file needs to contain the user name and password of a
highly privileged
.SM LDAP
user, so keeping it secure and secret is imperative for system security.
.SH EXAMPLES
A command line like
.IP
member_creator \-\-config=/etc/member_creator.conf
.PP
will run
.B member_creator
following the settings specified in the /etc/member_creator.conf configuration
file.
.SH FILES
.B member_creator
reads the configuration file specified as
.IR \-\-config .
No other files are used.
.TP
.B member_creator.conf
is the configuration file specified as
.IR \-\-config .
Its structure is described under
.I CONFIGURATION FILE
below.
.SH "CONFIGURATION FILE"
.PP
The
.B member_creator
configuration file consists of a number of sections which contain
.IR key :
.I value
pairs.
.PP
The following sections are recognized:
.SS database_config
This required section indicates where to find the
.IR cassandra (8)
database service used for finding and storing membership data.
.TP
.BI database_server " optional
.I host:port
pair to connect to for the
.IR cassandra (8)
database server.
There should be a database server listening to this port.
.IR default: " localhost:9160
.TP
.BI database_name " optional
.IR cassandra (8)
keyspace to bind to in order to find the relevant membership data.
It is expected that this keyspace has been initialized with the data structures
defined in the
.I cassandra\-schema
file accompanying the
.B membersys
source code.
.IR default: " sfmembersys
.TP
.BI database_timeout " optional
How long, in milliseconds, to wait for responses from the
.IR cassandra (8)
service before declaring a timeout error.
A value of 0 means to wait without any timeouts.
.IR default: " 0
.SS ldap_config
This required section contains the necessary configuration for porperly
accessing the
.SM LDAP
directory and managing users in it.
.TP
.BI server " required
.SM LDAP
URI pointing to a server holding the directory to work on.
.TP
.BI super_user " required
Prefix of an
.SM LDAP
account administrator user which will be used to modify the
.SM LDAP
tree.
The
.B base
configuration setting will be appended to this value after a comma.
.TP
.BI super_password " required
The bind password required to authenticate the
.BR super_user .
.TP
.BI base " required
.SM LDAP
base to append to all values.
Essentially, this would be the name of the toplevel
.SM LDAP
container.
.TP
.BI new_user_suffix " optional
Specifies a suffix to append to users (before
.BR base ).
This would be the name of the container which contains users in
.SM LDAP.
.IR default: " ou=People
.TP
.BI new_user_group " optional, repeated
Lists all groups which new members should be made a member of.
The groups are listed just as names, without the
.I cn=
prefix.
If a new member should be part of multiple groups (e.g. users and
.IR organization ),
just add multiple lines here.
.TP
.BI new_user_gid " required
POSIX group ID of the primary group the user should be a member of.
Obviously, this can only be specified once.
.TP
.BI new_user_shell " optional
This should be a path to a shell binary which should be used as the default
shell for new users.
.IR default: " /bin/bash
.SS welcome_mail_config
Optional settings to send welcome mails to newly created members.
If this section is ommitted no emails will be send.
.TP
.BI mail_template_path " required
Path to the template containing the welcome mail text.
.TP
.BI smtp_server_address " required
Name or address of the smtp server for outgoing mail.
.TP
.BI identity " optional
Set identity for SMTP authentication to act as a different role from username
(see RFC 4616).
Leave empty to use username instead.
.TP
.BI username " optional
Username for SMTP authentication.
If this and the password is specified authenticated SMTP will be used.
.TP
.BI password " optional
Password for SMTP authentication.
If this and the username is specified authenticated SMTP will be used.
.TP
.BI from " required
From header of the email.
.TP
.BI reply_to " optional
Reply-To header of the email.
Use this so new users reply to this address instead of the address specified in
from.
.TP
.BI subject " required
Subject header of the email.

.SH "EXAMPLE CONFIGURATION"
.PP
An example configuration file might look just about like this:
.PP
.RS 0
database_config {
.RS 0
}
.RS 0

.RS 0
ldap_config {
.RS 0
	server: "ldap.example.com"
.RS 0
	super_user: "cn=ldapadmin,ou=Users"
.RS 0
	super_password: "supersecret"
.RS 0
	base: "dc=example,dc=com"
.RS 0
	new_user_suffix: "ou=Users"
.RS 0
	new_user_group: "users"
.RS 0
	new_user_group: "exampleev"
.RS 0
	new_user_gid: 1024
.RS 0
	new_user_shell: "/sbin/nologin"
.RS 0
}
.RS 0

.RS 0
welcome_mail_config {
.RS 0
        mail_template_path: "/usr/share/membersys/mail.txt"
.RS 0
        smtp_server_address: "mail.example.com:587"
.RS 0
        username: "mailer@example.com"
.RS 0
        password: "secret"
.RS 0
        from: "Membership System <membersys@example.com>"
.RS 0
        reply_to: "<mailinglist@example.com>"
.RS 0
        subject: "Welcome to the example organization!"
.RS 0
}
.SH AUTHOR
Written by Caoimhe Chaos.
.SH "REPORTING BUGS"
Please report bugs via e-mail at
.I open\-at\-lists.starship-factory.ch
or file them via ditz and send a pull request to
.IR https://github.com/starshipfactory/membersys .
.PP
Please visit
.I http://bugs.starship-factory.ch/membersys/
for an overview of the current state of bug squashing.
.SH COPYRIGHT
.PP
Copyright \(co 2014 Caoimhe Chaos, Starship Factory. All Rights Reserved.
.PP
Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the conditions outlined in the
3-Clause BSD license are met.


ChangeLog
== 0.3 / 2014-09-02
* Keep members in the membership database too.
* Add a barcode for ease of finding the correct record to the print template
* Store the filled-out form data in a Cassandra database

== 0.2.1 / 2014-06-05
* bugfix: Fix print layout to fit onto a single page.
* bugfix: Reglement and statues don't link anywhere.

== 0.2 / 2014-02-09
* bugfix: Hide the print button from the final printed document
* Make it possible to have different prices for monthly and yearly payments
* Add a logo to the join website

== 0.1 / 2013-12-02
* Implement proof of concept web service which displays and accepts the membership form
* Integrate data and error messages into the application template
* Fill the data into the print template


member_creator/member_creator.go
/*
 * (c) 2014, Tonnerre Lombard <tonnerre@ancient-solutions.com>,
 *	     Starship Factory. All rights reserved.
 *
 * Redistribution and use in source  and binary forms, with or without
 * modification, are permitted  provided that the following conditions
 * are met:
 *
 * * Redistributions of  source code  must retain the  above copyright
 *   notice, this list of conditions and the following disclaimer.
 * * Redistributions in binary form must reproduce the above copyright
 *   notice, this  list of conditions and the  following disclaimer in
 *   the  documentation  and/or  other  materials  provided  with  the
 *   distribution.
 * * Neither  the name  of the Starship Factory  nor the  name  of its
 *   contributors may  be used to endorse or  promote products derived
 *   from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS"  AND ANY EXPRESS  OR IMPLIED WARRANTIES  OF MERCHANTABILITY
 * AND FITNESS  FOR A PARTICULAR  PURPOSE ARE DISCLAIMED. IN  NO EVENT
 * SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
 * INDIRECT, INCIDENTAL, SPECIAL,  EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED  TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE,  DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT  LIABILITY,  OR  TORT  (INCLUDING NEGLIGENCE  OR  OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package main

import (
	"code.google.com/p/goprotobuf/proto"
	"database/cassandra"
	"encoding/binary"
	"flag"
	"github.com/mqu/openldap"
	"io/ioutil"
	"log"
	"strconv"
	"strings"
	"time"
)

func makeMutation(mmap map[string][]*cassandra.Mutation, cf, name string,
	value []byte, now time.Time) {
	var m *cassandra.Mutation = cassandra.NewMutation()
	var col *cassandra.Column = cassandra.NewColumn()

	col.Name = []byte(name)
	col.Value = value
	col.Timestamp = now.UnixNano()

	m.ColumnOrSupercolumn = cassandra.NewColumnOrSuperColumn()
	m.ColumnOrSupercolumn.Column = col
	mmap[cf] = append(mmap[cf], m)
}

func makeMutationString(mmap map[string][]*cassandra.Mutation,
	cf, name, value string, now time.Time) {
	makeMutation(mmap, cf, name, []byte(value), now)
}

func makeMutationLong(mmap map[string][]*cassandra.Mutation,
	cf, name string, value uint64, now time.Time) {
	var val []byte = make([]byte, 8)
	binary.BigEndian.PutUint64(val, value)
	makeMutation(mmap, cf, name, val, now)
}

func makeMutationBool(mmap map[string][]*cassandra.Mutation,
	cf, name string, value bool, now time.Time) {
	var b byte
	if value {
		b = 0x1
	}
	makeMutation(mmap, cf, name, []byte{b}, now)
}

func asciiFilter(in string) string {
	var rv []rune
	var rn rune

	for _, rn = range []rune(in) {
		if rn < 128 {
			rv = append(rv, rn)
		}
	}

	return string(rv)
}

func main() {
	var cf string = "members"
	var config_file string
	var config_contents []byte
	var config MemberCreatorConfig
	var greatestUid uint64 = 1000
	var now time.Time
	var noop, verbose bool
        var welcome *WelcomeMail

	var ldap *openldap.Ldap
	var msg *openldap.LdapMessage
	var lres *openldap.LdapSearchResult

	var mmap map[string]map[string][]*cassandra.Mutation
	var db *cassandra.RetryCassandraClient
	var cp *cassandra.ColumnParent
	var pred *cassandra.SlicePredicate
	var kr *cassandra.KeyRange
	var kss []*cassandra.KeySlice
	var ks *cassandra.KeySlice

	var ire *cassandra.InvalidRequestException
	var ue *cassandra.UnavailableException
	var te *cassandra.TimedOutException
	var err error

	flag.StringVar(&config_file, "config", "",
		"Path to the member creator configuration file")
	flag.BoolVar(&noop, "dry-run", false, "Do a dry run")
	flag.BoolVar(&verbose, "verbose", false,
		"Whether or not to display verbose messages")
	flag.Parse()

	if len(config_file) == 0 {
		flag.Usage()
		return
	}

	config_contents, err = ioutil.ReadFile(config_file)
	if err != nil {
		log.Fatal("Unable to read ", config_file, ": ", err)
	}

	err = proto.Unmarshal(config_contents, &config)
	if err != nil {
		err = proto.UnmarshalText(string(config_contents), &config)
	}
	if err != nil {
		log.Fatal("Unable to parse ", config_file, ": ", err)
	}
        if config.WelcomeMailConfig != nil {
                welcome, err = NewWelcomeMail(config.WelcomeMailConfig)
                if err != nil {
                        log.Fatal("Error creating WelcomeMail: ", err)
                }
        }

	now = time.Now()

	if !noop {
		ldap, err = openldap.Initialize(config.LdapConfig.GetServer())
		if err != nil {
			log.Fatal("Error connecting to LDAP server ",
				config.LdapConfig.GetServer(), ": ", err)
		}

		err = ldap.SetOption(openldap.LDAP_OPT_PROTOCOL_VERSION,
			openldap.LDAP_VERSION3)
		if err != nil {
			log.Print("Error setting version to 3: ", err)
		}

		err = ldap.SetOption(openldap.LDAP_OPT_SIZELIMIT, 0)
		if err != nil {
			log.Print("Error setting the size limit to 0: ", err)
		}

		err = ldap.Bind(config.LdapConfig.GetSuperUser()+","+
			config.LdapConfig.GetBase(), config.LdapConfig.GetSuperPassword())
		if err != nil {
			log.Fatal("Unable to bind as ", config.LdapConfig.GetSuperUser()+
				","+config.LdapConfig.GetBase(), " to ",
				config.LdapConfig.GetServer(), ": ", err)
		}
		defer ldap.Unbind()

		// Find the highest assigned UID.
		msg, err = ldap.Search(config.LdapConfig.GetBase(),
			openldap.LDAP_SCOPE_SUBTREE, "(objectClass=posixAccount)",
			[]string{"uidNumber"})
		if err != nil {
			log.Fatal("Unable to search for posix accounts in ",
				config.LdapConfig.GetBase(), ": ", err)
		}
		for msg != nil {
			var entry = msg.FirstEntry()

			for entry != nil {
				var uid string

				for _, uid = range entry.GetValues("uidNumber") {
					var uidNumber uint64
					uidNumber, err = strconv.ParseUint(uid, 10, 64)
					if err != nil {
						log.Print("Error parsing \"", uid, "\" as a number: ",
							err)
					} else if uidNumber > greatestUid {
						greatestUid = uidNumber
					}
				}
				entry = entry.NextEntry()
			}

			msg = msg.NextMessage()
		}
	}

	// Connect to Cassandra so we can get a list of records to be processed.
	db, err = cassandra.NewRetryCassandraClient(
		config.DatabaseConfig.GetDatabaseServer())
	if err != nil {
		log.Fatal("Error connecting to Cassandra database at ",
			config.DatabaseConfig.GetDatabaseServer(), ": ", err)
	}

	ire, err = db.SetKeyspace(config.DatabaseConfig.GetDatabaseName())
	if ire != nil {
		log.Fatal("Invalid Cassandra request: ", ire.Why)
	}
	if err != nil {
		log.Fatal("Error setting keyspace: ", err)
	}

	cp = cassandra.NewColumnParent()
	cp.ColumnFamily = "membership_queue"
	pred = cassandra.NewSlicePredicate()
	pred.ColumnNames = [][]byte{[]byte("pb_data")}
	kr = cassandra.NewKeyRange()
	kr.StartKey = []byte("queue:")
	kr.EndKey = []byte("queue;")

	kss, ire, ue, te, err = db.GetRangeSlices(cp, pred, kr,
		cassandra.ConsistencyLevel_QUORUM)
	if ire != nil {
		log.Fatal("Invalid Cassandra request: ", ire.Why)
	}
	if ue != nil {
		log.Fatal("Cassandra unavailable")
	}
	if te != nil {
		log.Fatal("Cassandra timed out: ", te.String())
	}
	if err != nil {
		log.Fatal("Error getting range slice: ", err)
	}

	for _, ks = range kss {
		mmap = make(map[string]map[string][]*cassandra.Mutation)
		var csc *cassandra.ColumnOrSuperColumn
		for _, csc = range ks.Columns {
			var col *cassandra.Column = csc.Column
			var agreement MembershipAgreement
			var attrs map[string][]string
			var m *cassandra.Mutation

			if col == nil {
				continue
			}

			if string(col.Name) != "pb_data" {
				log.Print("Column selected was not as requested: ",
					col.Name)
				continue
			}

			err = proto.Unmarshal(col.Value, &agreement)
			if err != nil {
				log.Print("Unable to parse column ", col.Name, " of ",
					ks.Key, ": ", err)
				continue
			}

			greatestUid++
			attrs = make(map[string][]string)
			attrs["uidNumber"] = []string{
				strconv.FormatUint(greatestUid, 10)}
			attrs["gecos"] = []string{agreement.MemberData.GetName()}
			attrs["shadowLastChange"] = []string{"11457"}
			attrs["shadowMax"] = []string{"9999"}
			attrs["shadowWarning"] = []string{"7"}
			attrs["gidNumber"] = []string{strconv.FormatUint(uint64(
				config.LdapConfig.GetNewUserGid()), 10)}
			attrs["objectClass"] = []string{
				"account", "posixAccount", "shadowAccount", "top",
			}
			attrs["uid"] = []string{agreement.MemberData.GetUsername()}
			attrs["cn"] = []string{agreement.MemberData.GetUsername()}
			attrs["homeDirectory"] = []string{
				"/home/" + agreement.MemberData.GetUsername()}
			attrs["loginShell"] = []string{
				config.LdapConfig.GetNewUserShell()}
			attrs["userPassword"] = []string{
				agreement.MemberData.GetPwhash(),
			}

			for key, ivalue := range attrs {
				for i, value := range ivalue {
					attrs[key][i] = asciiFilter(value)
				}
			}

                        agreement.MemberData.Id = proto.Uint64(greatestUid)
                        col.Value, err = proto.Marshal(&agreement)
                        if err != nil {
                                log.Print("Error marshalling agreement: ", err)
                                continue
                        }


			if verbose {
				log.Print("Creating user: uid=" +
					agreement.MemberData.GetUsername() +
					"," + config.LdapConfig.GetNewUserSuffix() + "," +
					config.LdapConfig.GetBase())
			}

			if agreement.MemberData.Username == nil {
				if verbose {
					log.Print("Member without user account, nothing to do")
				}
			} else if noop {
				log.Print("Would create user: ", attrs)
			} else {
				var group string

				err = ldap.Add("uid="+agreement.MemberData.GetUsername()+
					","+config.LdapConfig.GetNewUserSuffix()+","+
					config.LdapConfig.GetBase(), attrs)
				if err != nil {
					log.Print("Unable to create user ",
						agreement.MemberData.GetUsername(), ": ", err)
					continue
				}

				attrs = make(map[string][]string)
				attrs["memberUid"] = []string{
					agreement.MemberData.GetUsername()}

				for _, group = range config.LdapConfig.GetNewUserGroup() {
					err = ldap.ModifyAdd("cn="+group+
						",ou=Groups,"+
						config.LdapConfig.GetBase(),
						attrs)
					if err != nil {
						log.Print("Unable to add user ",
							agreement.MemberData.GetUsername(),
							" to group ", group, ": ",
							err)
					}
				}
			}

			mmap["member:"+string(agreement.MemberData.GetEmail())] =
				make(map[string][]*cassandra.Mutation)
			mmap["member:"+string(agreement.MemberData.GetEmail())][cf] =
				make([]*cassandra.Mutation, 0)

			makeMutation(mmap["member:"+string(agreement.MemberData.GetEmail())],
				cf, "pb_data", col.Value, now)
			makeMutationString(mmap["member:"+string(agreement.MemberData.GetEmail())],
				cf, "name", agreement.MemberData.GetName(), now)
			makeMutationString(mmap["member:"+string(agreement.MemberData.GetEmail())],
				cf, "street", agreement.MemberData.GetStreet(), now)
			makeMutationString(mmap["member:"+string(agreement.MemberData.GetEmail())],
				cf, "city", agreement.MemberData.GetCity(), now)
			makeMutationString(mmap["member:"+string(agreement.MemberData.GetEmail())],
				cf, "country", agreement.MemberData.GetCountry(), now)
			makeMutationString(mmap["member:"+string(agreement.MemberData.GetEmail())],
				cf, "email", agreement.MemberData.GetEmail(), now)
			if agreement.MemberData.Phone != nil {
				makeMutationString(mmap["member:"+string(agreement.MemberData.GetEmail())],
					cf, "phone", agreement.MemberData.GetPhone(), now)
			}
			if agreement.MemberData.Username != nil {
				makeMutationString(mmap["member:"+string(agreement.MemberData.GetEmail())],
					cf, "username", agreement.MemberData.GetUsername(), now)
			}
			makeMutationLong(mmap["member:"+string(agreement.MemberData.GetEmail())],
				cf, "fee", agreement.MemberData.GetFee(), now)
			makeMutationBool(mmap["member:"+string(agreement.MemberData.GetEmail())],
				cf, "fee_yearly", agreement.MemberData.GetFeeYearly(), now)
			makeMutationLong(mmap["member:"+string(agreement.MemberData.GetEmail())],
				cf, "approval_ts", agreement.Metadata.GetApprovalTimestamp(),
				now)
			makeMutation(mmap["member:"+string(agreement.MemberData.GetEmail())],
				cf, "agreement_pdf", agreement.AgreementPdf, now)

			// Now, delete the original record.
			m = cassandra.NewMutation()
			m.Deletion = cassandra.NewDeletion()
			m.Deletion.Predicate = cassandra.NewSlicePredicate()
			m.Deletion.Predicate.ColumnNames = [][]byte{[]byte("pb_data")}
			m.Deletion.Timestamp = col.Timestamp

			mmap[string(ks.Key)] = make(map[string][]*cassandra.Mutation)
			mmap[string(ks.Key)]["membership_queue"] = []*cassandra.Mutation{m}

                        // Write welcome e-mail to new member.
                        if welcome != nil {
                                err = welcome.SendMail(agreement.MemberData)
                                if err != nil {
                                        log.Print("Error sending welcome e-mail to ",
                                        agreement.MemberData.GetEmail(), ": ", err)
                                }
                        }
		}

		// Apply all database mutations.
		ire, ue, te, err = db.BatchMutate(mmap, cassandra.ConsistencyLevel_QUORUM)
		if ire != nil {
			log.Fatal("Invalid Cassandra request: ", ire.Why)
		}
		if ue != nil {
			log.Fatal("Cassandra unavailable")
		}
		if te != nil {
			log.Fatal("Cassandra timed out: ", te.String())
		}
		if err != nil {
			log.Fatal("Error getting range slice: ", err)
		}
	}

	// Delete parting members.
	cp.ColumnFamily = "membership_dequeue"
	kr.StartKey = []byte("dequeue:")
	kr.EndKey = []byte("dequeue;")
	kss, ire, ue, te, err = db.GetRangeSlices(cp, pred, kr,
		cassandra.ConsistencyLevel_QUORUM)
	if ire != nil {
		log.Fatal("Invalid Cassandra request: ", ire.Why)
	}
	if ue != nil {
		log.Fatal("Cassandra unavailable")
	}
	if te != nil {
		log.Fatal("Cassandra timed out: ", te.String())
	}
	if err != nil {
		log.Fatal("Error getting range slice: ", err)
	}

	for _, ks = range kss {
		var csc *cassandra.ColumnOrSuperColumn
		mmap = make(map[string]map[string][]*cassandra.Mutation)
		for _, csc = range ks.Columns {
			var uuid cassandra.UUID
			var ldapuser string
			var col *cassandra.Column = csc.Column
			var agreement MembershipAgreement
			var attrs map[string][]string
			var m *cassandra.Mutation

			if col == nil {
				continue
			}

			if string(col.Name) != "pb_data" {
				log.Print("Column selected was not as requested: ",
					col.Name)
				continue
			}

			err = proto.Unmarshal(col.Value, &agreement)
			if err != nil {
				log.Print("Unable to parse column ", ks.Key, ": ", err)
				continue
			}

			ldapuser = "uid=" + agreement.MemberData.GetUsername() +
				"," + config.LdapConfig.GetNewUserSuffix() + "," +
				config.LdapConfig.GetBase()
			if agreement.MemberData.Username == nil {
				if verbose {
					log.Print("Member without user name, nothing to do")
				}
			} else if noop {
				log.Print("Would remove user ", ldapuser)
			} else {
				var groups []string
				var groups_differ bool
				var entry openldap.LdapEntry
				var group string

				lres, err = ldap.SearchAll(
					config.LdapConfig.GetBase(),
					openldap.LDAP_SCOPE_SUB,
					"(&(objectClass=posixGroup)(memberUid="+
						agreement.MemberData.GetUsername()+"))",
					[]string{"cn"})
				if err != nil {
					log.Print("Error searching groups for user ",
						agreement.MemberData.GetUsername(), ": ", err)
					continue
				}

				for _, entry = range lres.Entries() {
					var cn string
					var found bool

					cn, err = entry.GetOneValueByName("cn")
					if err != nil {
						log.Fatal("Error determining groups for ",
							agreement.MemberData.GetUsername(), ": ",
							err)
					}
					groups = append(groups, cn)

					for _, group = range config.LdapConfig.GetNewUserGroup() {
						if cn == group {
							found = true
						}
					}

					if !found {
						groups_differ = true
					}
				}

				if groups_differ {
					log.Print("User is in other groups than expected: ",
						strings.Join(groups, ", "))

					attrs = make(map[string][]string)
					attrs["memberUid"] = []string{
						agreement.MemberData.GetUsername()}
					for _, group = range config.LdapConfig.GetNewUserGroup() {
						ldap.ModifyDel("cn="+group+",ou=Groups,"+
							config.LdapConfig.GetBase(), attrs)
					}
				} else {
					// The user appears to be only in the groups given in
					// the config.
					err = ldap.Delete(ldapuser)
					if err != nil {
						log.Print("Unable to delete user ", ldapuser, ": ",
							err)
					}
				}
			}

			m = cassandra.NewMutation()
			m.Deletion = cassandra.NewDeletion()
			m.Deletion.Predicate = pred
			m.Deletion.Timestamp = col.Timestamp

			mmap[string(ks.Key)] = make(map[string][]*cassandra.Mutation)
			mmap[string(ks.Key)]["membership_dequeue"] = []*cassandra.Mutation{m}

			// 2 years retention.
			col.Ttl = 720 * 24 * 3600
			col.Timestamp = now.UnixNano()

			uuid = cassandra.UUIDFromBytes(ks.Key[len("dequeue:"):])

			m = cassandra.NewMutation()
			m.ColumnOrSupercolumn = cassandra.NewColumnOrSuperColumn()
			m.ColumnOrSupercolumn.Column = col
			mmap["archive:"+string([]byte(uuid))] = make(map[string][]*cassandra.Mutation)
			mmap["archive:"+string([]byte(uuid))]["membership_archive"] =
				[]*cassandra.Mutation{m}
		}

		// Apply all database mutations.
		ire, ue, te, err = db.BatchMutate(mmap, cassandra.ConsistencyLevel_QUORUM)
		if ire != nil {
			log.Fatal("Invalid Cassandra request: ", ire.Why)
		}
		if ue != nil {
			log.Fatal("Cassandra unavailable")
		}
		if te != nil {
			log.Fatal("Cassandra timed out: ", te.String())
		}
		if err != nil {
			log.Fatal("Error getting range slice: ", err)
		}
	}

	if verbose {
		log.Print("Greatest UID: ", greatestUid)
	}
}
